README.md
# minitest mock expectations

Provides method call assertions for minitest.

## Installation

Add this line to your application's Gemfile:

```ruby
gem "minitest-mock_expectations"
```

And then execute:

```bash
$ bundle
```

Or install it as:

```bash
$ gem install minitest-mock_expectations
```

## Usage

```ruby
require "minitest/mock_expectations"
```

Imagine we have model `Post`:

```ruby
class Post
  attr_accessor :title, :body
  attr_reader :comments

  def initialize(title: "", body: "", comments: [])
    @title = title
    @body = body
    @comments = comments
  end

  def add_comment(comment)
    @comments << comment

    "Thank you!"
  end
end
```

and `@post` variable that refers to an instance of `Post`:

```ruby
def setup
  @post = Post.new(
    title: "What is new in Rails 6.0",
    body: "https://gitlab.com/bogdanvlviv/posts/-/issues/15",
    comments: [
      "Wow.",
      "I like this post."
    ]
  )
end
```

### assert_called(object, method_name, message = nil, times: 1, returns: nil)

Asserts that the method will be called on the `object` in the block

```ruby
assert_called(@post, :title) do
  @post.title
end
```

To assert that the method will be called multiple times on the `object` in the block set `:times` option:

```ruby
assert_called(@post, :title, times: 2) do
  @post.title
  @post.title
end
```

You can stub the return value of the method in the block via `:returns` option:

```ruby
assert_called(@post, :title, returns: "What is new in Rails 5.2") do
  assert_equal "What is new in Rails 5.2", @object.title
end

assert_equal "What is new in Rails 6.0", @object.title
```

### refute_called(object, method_name, message = nil, &block)

Asserts that the method will not be called on the `object` in the block

```ruby
refute_called(@post, :title) do
  @post.body
end
```

### assert_not_called

Alias for `refute_called`.

### assert_called_with(object, method_name, arguments, returns: nil)

Asserts that the method will be called with the `arguments` on the `object` in the block

```ruby
assert_called_with(@post, :add_comment, ["Thanks for sharing this."]) do
  @post.add_comment("Thanks for sharing this.")
end
```

You can stub the return value of the method in the block via `:returns` option:

```ruby
assert_called_with(@post, :add_comment, ["Thanks for sharing this."], returns: "Thanks!") do
  assert_equal "Thanks!", @post.add_comment("Thanks for sharing this.")
end

assert_equal "Thank you!", @post.add_comment("Thanks for sharing this.")
```

You can also assert that the method will be called with different `arguments` on the `object` in the block:

```ruby
assert_called_with(@post, :add_comment, [["Thanks for sharing this."], ["Thanks!"]]) do
  @post.add_comment("Thanks for sharing this.")
  @post.add_comment("Thanks!")
end
```

```ruby
assert_called_with(@post, :add_comment, [[["Thanks for sharing this."]]]) do
  @post.add_comment(["Thanks for sharing this."])
end
```

```ruby
assert_called_with(@post, :add_comment, [[["Thanks for sharing this.", "Thanks!"]]]) do
  @post.add_comment(["Thanks for sharing this.", "Thanks!"])
end
```

```ruby
assert_called_with(@post, :add_comment, [[["Thanks for sharing this."], ["Thanks!"]]]) do
  @post.add_comment(["Thanks for sharing this."], ["Thanks!"])
end
```

```ruby
assert_called_with(@post, :add_comment, [["Thanks for sharing this."], {body: "Thanks!"}]) do
  @post.add_comment(["Thanks for sharing this."], {body: "Thanks!"})
end
```

```ruby
assert_called_with(@post, :add_comment, [[["Thanks for sharing this."]], [{body: "Thanks!"}]]) do
  @post.add_comment(["Thanks for sharing this."])
  @post.add_comment({body: "Thanks!"})
end
```

```ruby
assert_called_with(@post, :add_comment, [[["Thanks for sharing this."]], [["Thanks!"]]]) do
  @post.add_comment(["Thanks for sharing this."])
  @post.add_comment(["Thanks!"])
end
```

### assert_called_on_instance_of(klass, method_name, message = nil, times: 1, returns: nil)

Asserts that the method will be called on an instance of the `klass` in the block

```ruby
assert_called_on_instance_of(Post, :title) do
  @post.title
end
```

To assert that the method will be called multiple times on an instance of the `klass` in the block set `:times` option:

```ruby
assert_called_on_instance_of(Post, :title, times: 2) do
  @post.title
  @post.title
end
```

You can stub the return value of the method in the block via `:returns` option:

```ruby
assert_called_on_instance_of(Post, :title, returns: "What is new in Rails 5.2") do
  assert_equal "What is new in Rails 5.2", @post.title
end

assert_equal "What is new in Rails 6.0", @post.title
```

Use nesting of the blocks to assert that the several methods will be called on an instance of the `klass` in the block:

```ruby
assert_called_on_instance_of(Post, :title, times: 3) do
  assert_called_on_instance_of(Post, :body, times: 2) do
    @post.title
    @post.body
    @post.title
    @post.body
    @post.title
  end
end
```

### refute_called_on_instance_of(klass, method_name, message = nil, &block)

Asserts that the method will not be called on an instance of the `klass` in the block

```ruby
refute_called_on_instance_of(Post, :title) do
  @post.body
end
```

Use nesting of the blocks to assert that the several methods will not be called on an instance of the `klass` in the block:

```ruby
refute_called_on_instance_of(Post, :title) do
  refute_called_on_instance_of(Post, :body) do
    @post.add_comment("Thanks for sharing this.")
  end
end
```

### assert_not_called_on_instance_of

Alias for `refute_called_on_instance_of`.

## Contributing

Bug reports and pull requests are welcome on GitHub at https://github.com/bogdanvlviv/minitest-mock_expectations. This project is intended to be a safe, welcoming space for collaboration, and contributors are expected to adhere to the [Contributor Covenant](https://contributor-covenant.org) code of conduct.

## License

The gem is available as open source under the terms of the [MIT License](https://opensource.org/licenses/MIT).


CHANGELOG.md
### 1.2.0 / January 7, 2023

* Drop support for Ruby < 3.0.

### 1.1.3 / April 24, 2020

* Revert "Fix `assert_called_with` with an array as expected argument.".

### 1.1.2 / April 24, 2020

* Fix `assert_called_with` with an array as expected argument.

* Fix COPYRIGHT HOLDER of LICENSE.txt.

### 1.1.1 / February 26, 2019

### 1.1.0 / February 26, 2019

* Drop support for Ruby < 2.5.

### 1.0.0 / October 19, 2018

* [Birthday!](https://gitlab.com/bogdanvlviv/posts/-/issues/16)


test/test_helper.rb
$LOAD_PATH.unshift File.expand_path("../lib", __dir__)
require "minitest/mock_expectations"

require "minitest/autorun"


test/minitest/mock_expectations_test.rb
require "test_helper"

class Minitest::MockExpectationsTest < Minitest::Test
  def test_that_it_has_a_version_number
    refute_nil ::Minitest::MockExpectations::VERSION
  end
end


test/minitest/mock_expectations/assertions_test.rb
require "test_helper"

class Minitest::MockExpectations::AssertionsTest < Minitest::Test
  class Post
    attr_accessor :title, :body
    attr_reader :comments

    def initialize(title: "", body: "", comments: [])
      @title = title
      @body = body
      @comments = comments
    end

    def add_comment(comment)
      @comments << comment

      "Thank you!"
    end
  end

  def setup
    @post = Post.new(
      title: "What is new in Rails 6.0",
      body: "https://gitlab.com/bogdanvlviv/posts/-/issues/15",
      comments: [
        "Wow.",
        "I like this post."
      ]
    )
  end

  def test_assert_called_with_defaults_to_expect_once
    assert_called(@post, :title) do
      @post.title
    end
  end

  def test_assert_called_more_than_once
    assert_called(@post, :title, times: 2) do
      @post.title
      @post.title
    end
  end

  def test_assert_called_method_with_arguments
    assert_called(@post, :add_comment) do
      @post.add_comment("Thanks for sharing this.")
    end
  end

  def test_assert_called_returns
    assert_called(@post, :title, returns: "What is new in Rails 5.2") do
      assert_equal "What is new in Rails 5.2", @post.title
    end

    assert_equal "What is new in Rails 6.0", @post.title
  end

  def test_assert_called_failure_because_method_has_not_been_called
    error = assert_raises(Minitest::Assertion) do
      assert_called(@post, :title) do
        @post.body
      end
    end

    assert_equal "Expected title to be called 1 times, but was called 0 times.\nExpected: 1\n  Actual: 0", error.message
  end

  def test_assert_called_failure_because_method_has_been_called_more_than_expected_times
    error = assert_raises(Minitest::Assertion) do
      assert_called(@post, :title) do
        @post.title
        @post.title
      end
    end

    assert_equal "Expected title to be called 1 times, but was called 2 times.\nExpected: 1\n  Actual: 2", error.message
  end

  def test_assert_called_failure_with_message
    error = assert_raises(Minitest::Assertion) do
      assert_called(@post, :title, "Message") do
        @post.body
      end
    end

    assert_equal "Message.\nExpected title to be called 1 times, but was called 0 times.\nExpected: 1\n  Actual: 0", error.message
  end

  def test_refute_called
    refute_called(@post, :title) do
      @post.body
    end
  end

  def test_refute_called_failure_because_method_has_been_called
    error = assert_raises(Minitest::Assertion) do
      refute_called(@post, :add_comment) do
        @post.add_comment("Thanks for sharing this.")
      end
    end

    assert_equal "Expected add_comment to be called 0 times, but was called 1 times.\nExpected: 0\n  Actual: 1", error.message
  end

  def test_refute_called_failure_with_message
    error = assert_raises(Minitest::Assertion) do
      refute_called(@post, :title, "Message") do
        @post.title
      end
    end

    assert_equal "Message.\nExpected title to be called 0 times, but was called 1 times.\nExpected: 0\n  Actual: 1", error.message
  end

  def test_assert_not_called_is_alias_for_refute_called
    assert method(:assert_not_called).eql?(method(:refute_called))
  end

  def test_assert_called_with_arguments
    assert_called_with(@post, :add_comment, ["Thanks for sharing this."]) do
      @post.add_comment("Thanks for sharing this.")
    end
  end

  def test_assert_called_with_arguments_and_returns
    assert_called_with(@post, :add_comment, ["Thanks for sharing this."], returns: "Thanks!") do
      assert_equal "Thanks!", @post.add_comment("Thanks for sharing this.")
    end

    assert_equal "Thank you!", @post.add_comment("Thanks for sharing this.")
  end

  def test_assert_called_with_failure_because_method_has_not_been_called_with_expected_arguments
    assert_raises(MockExpectationError) do
      assert_called_with(@post, :add_comment, ["Thanks for sharing this."]) do
        @post.add_comment("Thanks")
      end
    end
  end

  def test_assert_called_with_multiple_expected_arguments
    assert_called_with(@post, :add_comment, [["Thanks for sharing this."], ["Thanks!"]]) do
      @post.add_comment("Thanks for sharing this.")
      @post.add_comment("Thanks!")
    end
  end

  def test_assert_called_with_an_array_as_expected_argument
    assert_called_with(@post, :add_comment, [[["Thanks for sharing this."]]]) do
      @post.add_comment(["Thanks for sharing this."])
    end

    assert_called_with(@post, :add_comment, [[["Thanks for sharing this.", "Thanks!"]]]) do
      @post.add_comment(["Thanks for sharing this.", "Thanks!"])
    end
  end

  def test_assert_called_with_expected_arguments_as_arrays
    assert_called_with(@post, :add_comment, [[["Thanks for sharing this."], ["Thanks!"]]]) do
      @post.add_comment(["Thanks for sharing this."], ["Thanks!"])
    end
  end

  def test_assert_called_with_multiple_expected_arguments_as_arrays
    assert_called_with(@post, :add_comment, [[["Thanks for sharing this."]], [["Thanks!"]]]) do
      @post.add_comment(["Thanks for sharing this."])
      @post.add_comment(["Thanks!"])
    end
  end

  def test_assert_called_with_expected_arguments_as_array_and_one_non_array_object
    assert_called_with(@post, :add_comment, [["Thanks for sharing this."], {body: "Thanks!"}]) do
      @post.add_comment(["Thanks for sharing this."], {body: "Thanks!"})
    end
  end

  def test_assert_called_with_multiple_expected_arguments_as_array_and_one_non_array_object
    assert_called_with(@post, :add_comment, [[["Thanks for sharing this."]], [{body: "Thanks!"}]]) do
      @post.add_comment(["Thanks for sharing this."])
      @post.add_comment({body: "Thanks!"})
    end
  end

  def test_assert_called_on_instance_of_with_defaults_to_expect_once
    assert_called_on_instance_of(Post, :title) do
      @post.title
    end
  end

  def test_assert_called_on_instance_of_more_than_once
    assert_called_on_instance_of(Post, :title, times: 2) do
      @post.title
      @post.title
    end
  end

  def test_assert_called_on_instance_of_with_arguments
    assert_called_on_instance_of(Post, :add_comment) do
      @post.add_comment("Thanks for sharing this.")
    end
  end

  def test_assert_called_on_instance_of_returns
    assert_called_on_instance_of(Post, :title, returns: "What is new in Rails 5.2") do
      assert_equal "What is new in Rails 5.2", @post.title
    end

    assert_equal "What is new in Rails 6.0", @post.title
  end

  def test_assert_called_on_instance_of_failure_because_method_has_not_been_called
    error = assert_raises(Minitest::Assertion) do
      assert_called_on_instance_of(Post, :title) do
        @post.body
      end
    end

    assert_equal "Expected title to be called 1 times, but was called 0 times.\nExpected: 1\n  Actual: 0", error.message
  end

  def test_assert_called_on_instance_of_failure_because_method_has_been_called_more_than_expected_times
    error = assert_raises(Minitest::Assertion) do
      assert_called_on_instance_of(Post, :title) do
        @post.title
        @post.title
      end
    end

    assert_equal "Expected title to be called 1 times, but was called 2 times.\nExpected: 1\n  Actual: 2", error.message
  end

  def test_assert_called_on_instance_of_failure_with_message
    error = assert_raises(Minitest::Assertion) do
      assert_called_on_instance_of(Post, :title, "Message") do
        @post.body
      end
    end

    assert_equal "Message.\nExpected title to be called 1 times, but was called 0 times.\nExpected: 1\n  Actual: 0", error.message
  end

  def test_assert_called_on_instance_of_nesting
    assert_called_on_instance_of(Post, :title, times: 3) do
      assert_called_on_instance_of(Post, :body, times: 2) do
        @post.title
        @post.body
        @post.title
        @post.body
        @post.title
      end
    end
  end

  def test_refute_called_on_instance_of
    refute_called_on_instance_of(Post, :title) do
      @post.body
    end
  end

  def test_refute_called_on_instance_of_failure_because_method_has_been_called
    error = assert_raises(Minitest::Assertion) do
      refute_called_on_instance_of(Post, :add_comment) do
        @post.add_comment("Thanks for sharing this.")
      end
    end

    assert_equal "Expected add_comment to be called 0 times, but was called 1 times.\nExpected: 0\n  Actual: 1", error.message
  end

  def test_refute_called_on_instance_of_failure_with_message
    error = assert_raises(Minitest::Assertion) do
      refute_called_on_instance_of(Post, :title, "Message") do
        @post.title
      end
    end

    assert_equal "Message.\nExpected title to be called 0 times, but was called 1 times.\nExpected: 0\n  Actual: 1", error.message
  end

  def test_refute_called_on_instance_of_nesting
    refute_called_on_instance_of(Post, :title) do
      refute_called_on_instance_of(Post, :body) do
        @post.add_comment("Thanks for sharing this.")
      end
    end
  end

  def test_assert_not_called_on_instance_of_is_alias_for_refute_called
    assert method(:assert_not_called_on_instance_of).eql?(method(:refute_called_on_instance_of))
  end
end


lib/minitest/mock_expectations/assertions.rb
require "minitest"
require "minitest/assertions"
require "minitest/mock"

module Minitest
  # Provides method call assertions for minitest
  #
  # Imagine we have model +Post+:
  #
  #   class Post
  #     attr_accessor :title, :body
  #     attr_reader :comments
  #
  #     def initialize(title: "", body: "", comments: [])
  #       @title = title
  #       @body = body
  #       @comments = comments
  #     end
  #
  #     def add_comment(comment)
  #       @comments << comment
  #
  #       "Thank you!"
  #     end
  #   end
  #
  # and variable +@post+ that reffers to instance of +Post+:
  #
  #   def setup
  #     @post = Post.new(
  #       title: "What is new in Rails 6.0",
  #       body: "https://gitlab.com/bogdanvlviv/posts/-/issues/15",
  #       comments: [
  #         "Wow.",
  #         "I like this post."
  #       ]
  #     )
  #   end
  module Assertions
    # Asserts that the method will be called on the +object+ in the block
    #
    #   assert_called(@post, :title) do
    #     @post.title
    #   end
    #
    # In order to assert that the method will be called multiple times on the +object+ in the block set +:times+ option:
    #
    #   assert_called(@post, :title, times: 2) do
    #     @post.title
    #     @post.title
    #   end
    #
    # You can stub the return value of the method in the block via +:returns+ option:
    #
    #   assert_called(@post, :title, returns: "What is new in Rails 5.2") do
    #     assert_equal "What is new in Rails 5.2", @object.title
    #   end
    #
    #   assert_equal "What is new in Rails 6.0", @object.title
    def assert_called(object, method_name, message = nil, times: 1, returns: nil)
      times_called = 0

      object.stub(method_name, proc { times_called += 1; returns }) { yield }

      error = "Expected #{method_name} to be called #{times} times, but was called #{times_called} times"
      error = "#{message}.\n#{error}" if message
      assert_equal times, times_called, error
    end

    # Asserts that the method will not be called on the +object+ in the block
    #
    #   refute_called(@post, :title) do
    #     @post.body
    #   end
    def refute_called(object, method_name, message = nil, &block)
      assert_called(object, method_name, message, times: 0, &block)
    end

    alias assert_not_called refute_called

    # Asserts that the method will be called with the +arguments+ on the +object+ in the block
    #
    #   assert_called_with(@post, :add_comment, ["Thanks for sharing this."]) do
    #     @post.add_comment("Thanks for sharing this.")
    #   end
    #
    # You can stub the return value of the method in the block via +:returns+ option:
    #
    #   assert_called_with(@post, :add_comment, ["Thanks for sharing this."], returns: "Thanks!") do
    #     assert_equal "Thanks!", @post.add_comment("Thanks for sharing this.")
    #   end
    #
    #   assert_equal "Thank you!", @post.add_comment("Thanks for sharing this.")
    #
    # You can also assert that the method will be called with different +arguments+ on the +object+ in the block:
    #
    #   assert_called_with(@post, :add_comment, [["Thanks for sharing this."], ["Thanks!"]]) do
    #     @post.add_comment("Thanks for sharing this.")
    #     @post.add_comment("Thanks!")
    #   end
    #
    #   assert_called_with(@post, :add_comment, [[["Thanks for sharing this."]]]) do
    #     @post.add_comment(["Thanks for sharing this."])
    #   end
    #
    #   assert_called_with(@post, :add_comment, [[["Thanks for sharing this.", "Thanks!"]]]) do
    #     @post.add_comment(["Thanks for sharing this.", "Thanks!"])
    #   end
    #
    #   assert_called_with(@post, :add_comment, [[["Thanks for sharing this."], ["Thanks!"]]]) do
    #     @post.add_comment(["Thanks for sharing this."], ["Thanks!"])
    #   end
    #
    #   assert_called_with(@post, :add_comment, [["Thanks for sharing this."], {body: "Thanks!"}]) do
    #     @post.add_comment(["Thanks for sharing this."], {body: "Thanks!"})
    #   end
    #
    #   assert_called_with(@post, :add_comment, [[["Thanks for sharing this."]], [{body: "Thanks!"}]]) do
    #     @post.add_comment(["Thanks for sharing this."])
    #     @post.add_comment({body: "Thanks!"})
    #   end
    #
    #   assert_called_with(@post, :add_comment, [[["Thanks for sharing this."]], [["Thanks!"]]]) do
    #     @post.add_comment(["Thanks for sharing this."])
    #     @post.add_comment(["Thanks!"])
    #   end
    #
    def assert_called_with(object, method_name, arguments, returns: nil)
      mock = Minitest::Mock.new

      if arguments.all? { |argument| argument.is_a?(Array) }
        arguments.each { |argument| mock.expect(:call, returns, argument) }
      else
        mock.expect(:call, returns, arguments)
      end

      object.stub(method_name, mock) { yield }

      mock.verify
    end

    # Asserts that the method will be called on an instance of the +klass+ in the block
    #
    #   assert_called_on_instance_of(Post, :title) do
    #     @post.title
    #   end
    #
    # In order to assert that the method will be called multiple times on an instance of the +klass+ in the block set +:times+ option:
    #
    #   assert_called_on_instance_of(Post, :title, times: 2) do
    #     @post.title
    #     @post.title
    #   end
    #
    # You can stub the return value of the method in the block via +:returns+ option:
    #
    #   assert_called_on_instance_of(Post, :title, returns: "What is new in Rails 5.2") do
    #     assert_equal "What is new in Rails 5.2", @post.title
    #   end
    #
    #   assert_equal "What is new in Rails 6.0", @post.title
    #
    # Use nesting of the blocks in order assert that the several methods will be called on an instance of the +klass+ in the block:
    #
    #   assert_called_on_instance_of(Post, :title, times: 3) do
    #     assert_called_on_instance_of(Post, :body, times: 2) do
    #       @post.title
    #       @post.body
    #       @post.title
    #       @post.body
    #       @post.title
    #     end
    #   end
    def assert_called_on_instance_of(klass, method_name, message = nil, times: 1, returns: nil)
      times_called = 0

      klass.define_method("stubbed_#{method_name}") do |*|
        times_called += 1

        returns
      end

      klass.alias_method("original_#{method_name}", method_name)
      klass.alias_method(method_name, "stubbed_#{method_name}")

      yield

      error = "Expected #{method_name} to be called #{times} times, but was called #{times_called} times"
      error = "#{message}.\n#{error}" if message

      assert_equal times, times_called, error
    ensure
      klass.alias_method(method_name, "original_#{method_name}")
      klass.undef_method("original_#{method_name}")
      klass.undef_method("stubbed_#{method_name}")
    end

    # Asserts that the method will not be called on an instance of the +klass+ in the block
    #
    #   refute_called_on_instance_of(Post, :title) do
    #     @post.body
    #   end
    #
    # Use nesting of the blocks in order assert that the several methods will not be called on an instance of the +klass+ in the block:
    #
    #   refute_called_on_instance_of(Post, :title) do
    #     refute_called_on_instance_of(Post, :body) do
    #       @post.add_comment("Thanks for sharing this.")
    #     end
    #   end
    def refute_called_on_instance_of(klass, method_name, message = nil, &block)
      assert_called_on_instance_of(klass, method_name, message, times: 0, &block)
    end

    alias assert_not_called_on_instance_of refute_called_on_instance_of
  end
end
